### ISSUE TITLE
Temperature readings inaccurate and flapping after release 1.3.2 

### ISSUE THREAD

**Issue by medlor** [lebuni/ZACwire-Library#12]
Hi Adrian,

hope you have a nice time in your study. A couple of days ago I found out that the temperature readings of your great library is flapping a couple of 1/10 of degrees every now and then. My pid control handles this after a couple of seconds but still this is rather ugly.
I took me some time to find our that there must be a bug in the lib after including v1.3.2.
Latest working version is 1.3.1.

I have added a simple sketch.
Running on ESP32 (have not tested esp8266)

Symptoms:
1. very rough readings.. eg. jumps from 26.50 directly to 28.10 (when pressing my finger on the sensor) -> see log below

Test still running on version 1.3.1 to determine random flapping also occur in this version (couple of 1/10 of degrees every now and then).

Sample output: 

```
1.3.2
12 Temp: 26.50
12 Temp: 26.50
13 Temp: 28.10 *** 1.60
13 Temp: 28.10
13 Temp: 28.10
13 Temp: 28.10
.....
18 Temp: 28.10
18 Temp: 28.10
18 Temp: 26.50 *** 1.60
18 Temp: 28.10 *** 1.60
18 Temp: 26.50 *** 1.60  ERR!!!!
18 Temp: 28.10 *** 1.60
18 Temp: 26.50 *** 1.60
18 Temp: 26.50
18 Temp: 26.50
```

```
1.3.1
5 Temp: 26.10
5 Temp: 26.10
5 Temp: 26.10
5 Temp: 26.10
5 Temp: 26.30   OK (all lines)
5 Temp: 26.40
6 Temp: 26.50
6 Temp: 26.60
6 Temp: 26.80
6 Temp: 26.80
6 Temp: 26.90
6 Temp: 27.00
6 Temp: 27.00
6 Temp: 27.20 *** 0.20
6 Temp: 27.20
6 Temp: 27.30
7 Temp: 27.40
7 Temp: 27.40
7 Temp: 27.60 *** 0.20
7 Temp: 27.60
7 Temp: 27.60
7 Temp: 27.50
7 Temp: 27.50
7 Temp: 27.40
7 Temp: 27.40
7 Temp: 27.40
8 Temp: 27.30
8 Temp: 27.30
8 Temp: 27.20
8 Temp: 27.20
8 Temp: 27.20
8 Temp: 27.10
```

best regards
Tobias

**Comment by medlor** [comment]
[platformio.zip](https://github.com/lebuni/ZACwire-Library/files/7145875/platformio.zip)

**Comment by medlor** [comment]
a bitwindow of 110 instead of 130 improves the situation. (ESP32, TSIC306)
Could it be that the parity checks are not working or the latest adaptions to the calculation are off a bit?

**Comment by medlor** [comment]
what the the recommended way to determine the best bitwindow with only using software?

**Comment by medlor** [comment]
**About the "version 1.3.1 to determine random flapping" issue:**

every one and then I got strange outliers. I improved the logging on bleeding-edge to better see what happens. The log is created using the fork https://github.com/medlor/ZACwire-Library with a bitwindow of 110.
```
validation is the latest temp from getTemp().
actual is the temperature 100ms before that (Iuse "validation" temp and "prev" to detect outliers).
prev is 100ms before "actual", which is already validated by software-means. You can assume this is the real temperature.

You can see that "validation temp" is flapping.
(D p:^4061ms) 941 temp sensor read unrealistic: consecErrors=1, validation=98.40, actual=92.10
(D p:^4874ms) 1127 temp sensor read unrealistic: consecErrors=1, validation=98.40, actual=92.10
(D p:^4974ms) 1142 temp sensor read unrealistic: consecErrors=1, validation=142.10, actual=92.10
(D p:^0990ms) 1998 temp sensor read unrealistic: consecErrors=1, validation=-15.40, actual=84.60
(E p:^2598ms) 2138 temp sensor inaccuracy (TEMPSENSOR_BITWINDOW?): prev=91.70, actual=88.50, validation=91.70
(D p:^1398ms) 2152 temp sensor read unrealistic: consecErrors=1, validation=79.30, actual=91.90
(E p:^3000ms) 2155 temp sensor inaccuracy (TEMPSENSOR_BITWINDOW?): prev=91.90, actual=88.70, validation=91.80
(D p:^3398ms) 2199 temp sensor read unrealistic: consecErrors=1, validation=-8.00, actual=92.10
(E p:^1498ms) 2332 temp sensor inaccuracy (TEMPSENSOR_BITWINDOW?): prev=92.00, actual=88.90, validation=92.00
(D p:^1098ms) 2377 temp sensor read unrealistic: consecErrors=1, validation=98.20, actual=92.00
(E p:^4998ms) 2400 temp sensor inaccuracy (TEMPSENSOR_BITWINDOW?): prev=92.00, actual=88.80, validation=92.00
(E p:^2495ms) 2433 temp sensor inaccuracy (TEMPSENSOR_BITWINDOW?): prev=91.90, actual=88.80, validation=92.00
(E p:^4995ms) 2465 temp sensor inaccuracy (TEMPSENSOR_BITWINDOW?): prev=91.90, actual=93.50, validation=92.00
(E p:^2095ms) 2803 temp sensor inaccuracy (TEMPSENSOR_BITWINDOW?): prev=92.10, actual=88.90, validation=92.10
(D p:^0395ms) 2946 temp sensor read unrealistic: consecErrors=1, validation=66.90, actual=91.90
(D p:^4795ms) 2990 temp sensor read unrealistic: consecErrors=1, validation=84.10, actual=92.00
```
It might be that this is related to the ESP32's interrupts not working as expected in certain situations (see https://www.espressif.com/sites/default/files/documentation/eco_and_workarounds_for_bugs_in_esp32_en.pdf and related issues)?! 
I have solved such issues by implementing a simple validation function, which compares the last 3 samples to detect outliers. This might not be advisable to implement in a generic function such as in your lib.

**Comment by medlor** [comment]
to sum it up:
* it would be great if you can check the parity and/or temp-calculation issue (first issue). v1.3.1 seems to work much better.
* regarding esp32's outliers: I am not sure you/we can find a solution here. It sounds like an hardware issues. Nevertheless this is solved in bleeding-edge software. Therefore not critical for me anymore.

**Comment by TheTrueAI** [comment]
I tried your test code on a TSIC 206 and a TSIC 306 with an ESP32-WROOM-32 and ESP8266 12-F. Both sensors were mounted inside machines with glue, so I couldn't press my finger on them. I just heated them to like 100°C and down again for an hour.
With the current release 1.3.3, no single flapping occured on both sensors/microcontrollers.

BTW the `defaultBitWindow` is just important for the first few minutes, afterwards the code adjusts itself in a more precise way. I added that clarification to the readme
